src/psyclone/gocean1p0.py

```diff
@@ -1549,98 +1528,397 @@ def resolve_argument_names(args):
-                    try:
-                        read_fp = symtab.lookup_with_tag("ocl_read_func").name
-                    except KeyError:
-                        # If the subroutines does not exist, it needs to be
-                        # generated first.
-                        read_fp = self.gen_ocl_read_from_device_function(
-                            parent.parent)
-                    ifthen.add(AssignGen(
-                        ifthen, lhs="{0}%read_from_device_f".format(arg.name),
-                        rhs=read_fp, pointer=True))
-
-                # Ensure data copies have finished
-                ifthen.add(CommentGen(
-                    ifthen, " Block until data copies have finished"))
-                ifthen.add(AssignGen(ifthen, lhs=flag,
-                                     rhs="clFinish(" + qlist + "(1))"))
+                    arguments.append(arg.name)
+            elif arg.argument_type == "field":
+                # Cast buffer to cl_mem type expected by OpenCL
+                field = symtab.lookup(arg.name)
+                symbol = symtab.lookup_with_tag(arg.name + "_cl_mem")
+                source = StructureReference.create(field, ['device_ptr'])
+                dest = Reference(symbol)
+                bop = BinaryOperation.create(BinaryOperation.Operator.CAST,
+                                             source, dest)
+                assig = Assignment.create(dest.copy(), bop)
+                parent.add(PSyIRGen(parent, assig))
+
+                arguments.append(arg.name + "_cl_mem")
+            elif arg.argument_type == "grid_property":
+                garg = self._arguments.find_grid_access()
+                if arg.is_scalar:
+                    arguments.append(arg.dereference(garg.name))
+                else:
+                    # TODO (dl_esm_inf/#18) the dl_esm_inf library stores
+                    # the pointers to device memory for grid properties in
+                    # "<grid-prop-name>_device" which is a bit hacky but
+                    # works for now.
+                    arguments.append(garg.name+"%grid%"+arg.name+"_device")
+
+        sub_name = symtab.lookup_with_tag(self.name + "_set_args").name
+        parent.add(CallGen(parent, sub_name, arguments))
+
+    def gen_ocl_write_grid_buffers(self, f2pygen_module):
+        '''
+        Insert a f2pygen subroutine to initialise the OpenCL buffer in a
+        OpenCL device using FortCL in the given f2pygen module and return
+        its name.
+
+        :param f2pygen_module: the module where the new function will be \
+                               inserted.
+        :param type: :py:class:`psyclone.f2pygen.ModuleGen`
+
+        :returns: the name of the generated subroutine.
+        :rtype: str
+
+        '''
+        symtab = self.root.symbol_table
+
+        # Create the symbol for the routine and add it to the symbol table.
+        subroutine_name = symtab.new_symbol(
```

Comment:
I refactored this and all similar methods to do the try/except of the tag lookup inside the method instead of being done my the caller before like before. This converts the method to "always return me this routine and create it if it doesn't exist already".

I think this is better now because the caller code is simpler and new_symbol can not fail with an used tag (because we just check it before)

Also it returns the symbol instead of the name string to make it more PSyIR-like style.

Reply:
Yes, I agree that that's nicer.